pkg/clusteragent/externalmetrics/provider.go

```diff
@@ -113,9 +113,6 @@ func (p *datadogMetricProvider) GetExternalMetric(namespace string, metricSelect
 func (p *datadogMetricProvider) getExternalMetric(namespace string, metricSelector labels.Selector, info provider.ExternalMetricInfo) (*external_metrics.ExternalMetricValueList, error) {
 	log.Debugf("Received external metric query with ns: %s, selector: %s, metricName: %s", namespace, metricSelector.String(), info.Metric)
 
-	// Convert metric name to lower case to allow proper matching (and DD metrics are always lower case)
-	info.Metric = strings.ToLower(info.Metric)
```

Comment:
Hey @sebbrandt87,

I further checked the current behaviour and in fact this won't solve your problem.

The `strings.ToLower` ensures we get lower case metric name, but in fact the HPA controller already sends the metricName as lowercase (so case is lost before reaching us), so it's more to replicate this behaviour in unit tests, etc.
You can verify that by activating DEBUG logs and checking the log line above.

What happens when you put an uppercase in your HPA's metricName is that an `autogen-` DatadogMetric is created keeping the case properly and the query made to Datadog keeps the case but this `autogen-` metric cannot be matched with incoming query due to this, leading a "not found" response from the Cluster Agent.

However, using a [custom DatadogMetric object](https://docs.datadoghq.com/agent/cluster_agent/external_metrics/?tab=helm#example-datadogmetric-object) allows to bypass this limitation and should fully support case correctly.

Reply:
> What happens when you put an uppercase in your HPA's metricName is that an autogen- DatadogMetric is created keeping the case properly and the query made to Datadog keeps the case but this autogen- metric cannot be matched with incoming query due to this, leading a "not found" response from the Cluster Agent.

And this is the reason this PR was created and the unit test is proving that it is working with this change. Isn't it?

Reply:
Yes, but it's not relevant since the function will always get a lower case input, regardless of what you put inside the HPA.

Reply:
This is a big migration issue for us. We need to enable CustomMetrics CRD and all our current HPA metrics pointed to a case sensitive custom metric break. As stated, custom metrics are allowed to be uppercase and so this fix here would match that behaviour.

Reply:
Hey @sebbrandt87,

I'll try to explain differently. Even we merge this change, it still won't work, because what is received by this function is always lowercase.
You may check that for yourself by activating `DEBUG` logs in the Cluster Agent and check for this log (line 114, so before `strings.ToLower` is called):
```
log.Debugf("Received external metric query with ns: %s, selector: %s, metricName: %s", namespace, metricSelector.String(), info.Metric)
```

You'll see that `info.Metric` is already lowercase, this is inherent to how the `ExternalMetrics` is implemented in Kubernetes, and impact any `ExternalMetric` provider, not only Datadog Cluster Agent, see https://github.com/kubernetes/kubernetes/issues/72996 for more explanation.

Reply:
Hey @vboulineau,
okay we might come to the real root cause 😄 . Could it be that [this](https://github.com/DataDog/datadog-agent/pull/3013) logic got removed in the meantime and so the implementation is broken because of that? It seems that I haven't seen recognized https://github.com/kubernetes/kubernetes/issues/72996. 👍

Reply:
If I request with the lowercased metric name `domain.concurrentrequests`

```bash
kubectl get --raw "/apis/external.metrics.k8s.io/v1beta1/namespaces/search/domain.concurrentrequests?labelSelector=env%3Dlive" | jq .
```

 I get this result

```json
{
      "metricName": "domain.concurrentRequests",
      "metricLabels": {
        "env": "live"
      },
      "timestamp": "2022-01-26T11:58:47Z",
      "value": "0"
    }
```

so it is 100% true, also related to kubernetes/kubernetes#72996 that we only can query the kubernetes extern.metrics.k8s.io API in this way.

Reply:
so what I did now, because I cannot get local unit tests to work at the moment, changed `queryMetricName: "domain.concurrentrequests",` to be lowercase.

Reply:
If you don't see the `DEBUG` logs above while `DEBUG` logs are activated, it means you are using the legacy implementation (without `DatadogMetrics` CRD at all), so you're no going through the code path you updated here, but to the legacy one: `pkg/clusteragent/custommetrics`, in which case it should works, so I'm not clear on the configuration you use.

There is a way we could support this in autogen `DatadogMetrics` but it's not straightforward and not easy to understand the limitations:
- If you have two metrics that are only different due to case (e.g. `FOO.bar` and `foo.bar`), one of the two will "win" (we cannot predict which one), and the "winner" value will silently be used for both.
- If you create an HPA with a wrong metric name (`Foo.bar` instead of `foo.bar` for instance), we'll need to change the Autogen metrics query while keeping the same id, which was not intended in the original design.
- If you have two HPAs referencing the same metrics+labels but you made a mistake in one (like `Foo.bar` and `foo.bar`), it will work 50% of the time (same value for both) and fail 50% of the time (no value for both).

While not common, these issues are surprising and difficult to debug.

These issues are present in the `legacy` implementation and it's impossible to fix them as we always get a lower case `metricName` from Kubernetes.
For these reasons I strongly recommend to go the safest way, which is to use `DatadogMetrics` option, which does not have any of these issues.

Reply:
I have understood the issue already @vboulineau 🙇 
I just try to implement, out of a customer perspective, the DataDog [documented](https://docs.datadoghq.com/metrics/custom_metrics/#naming-custom-metrics) behaviour and rule.

So the way to go is in my opinion: 
> If you create an HPA with a wrong metric name (Foo.bar instead of foo.bar for instance), we'll need to change the Autogen metrics query while keeping the same id, which was not intended in the original design.

I am happy to help with that. I would for now set this PR as WIP again and see in the next days how I find the time to proceed, if this is okay for you?

Reply:
I'll circle back internally on what's the best solution, but I'm not sure we want to change the current behaviour.

Reply:
any news here @vboulineau ?

Reply:
Hey @sebbrandt87,

We chose to keep the current implementation. We'll document the limitation on Autogen metrics and recommend to use `DatadogMetric` directly when using case sensitive metrics name.